src/cromshell/status/command.py

```diff
@@ -114,34 +105,50 @@ def main(config, workflow_id):
-
-        # For each call name add the number of execution_status to a list
-        for instanceDic in workflow_metadata["calls"][call]:
-            execution_statuses.append(instanceDic["executionStatus"])
-
-        # Create a key and value pair for the execution status and number of times
-        # it is seen
-        status_count = Counter(execution_statuses)
-        # For each status for the call add the call name and status name and count
-        # to list
-        for status in set(execution_statuses):
-            status_element = f'{{"{status}": {status_count[status]}}}'
-            tmp_execution_status.append(f"{call_element}{status_element}")
-
-    tmp_execution_status_json = ", ".join(tmp_execution_status)
-    tmp_execution_status_json = "{" + tmp_execution_status_json + "}"
-    return [json.loads(tmp_execution_status_json)]
+def check_for_failure(metadata: dict):
+    """Checks a workflow metadata dictionary for failing statuses
+    Returns True to indicate workflow or some task(s) has failed"""
+
+    # If the given dictionary contains a 'status' key and has value of "Failed"
+    # then exit the function returning "True" to indicate workflow has failed
+    if metadata.get("status") == "Failed":
+        return True
+
+    # If the dictionary does not contain a failed value for its status key or
+    # if status key does does not exist then we'll iterate through the dictionary
+    # value in search of a nested dictionary (indicating subworkflows)
+    # or a list which holds the task statuses.
+    for value in metadata.values():
+
+        # If a dictionary value is encountered then Depth First Search recursion
+        # is used to traverse the dictionary by iterating through the given dictionary
+        # and reporting whether it found a failure. The reason for this is due to the
+        # tree structure of the metadata dictionary which holds the task statuses
+        # in second layer dictionary key called calls.
+        if isinstance(value, dict):
+            if check_for_failure(value):
+                return True
+
+        # If a list value is encountered then the dictionary being traversed through is
+        # a key and value pair where the key is the task name and the value is the
+        # number of times the task was executed (shard), which is normally 1 unless the
+        # task was scattered. Hence the value for the key is a list to account
+        # for all the shards for a task. Each item(shard) within this list is a
+        # dictionary holding status for the shard.
+        # We'll want to check each shard to determine whether it's status has "Failed".
+        if isinstance(value, list):
+            for shard in value:
+
+                # If a key in the shared labeled "subWorkflowMetadata" then this
+                # will another dictionary layer of subworkflow tasks that needs
+                # to be traversed.
+                if "subWorkflowMetadata" in shard.keys():
```

Comment:
Is the shard guaranteed to be a dictionary?  If not, won't this be an issue if it doesn't have a `keys()` method?

Reply:
From the workflow metadata I've seen so far, yes. If we come across a key in a dictionary with a value type of list then each item in that list is always a dictionary. The items within the list represent the number of times the task was executed, it could be once or multiple items for a scattered job. Each item is dictionary which hold a key and value pair for informations like executionStatus, backendStatus, shardIndex, callRoot                          attempt.